arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2953 @@
+            }
+          else
+            {
+              pmsg->data_low |= (cf->data[i] << ((7 - i) * 8));
+            }
+        }
+    }
+
+  /* Insert CAN msg object into available TX bf */
+
+  if (CAN_VALID_MSG != (ret = mpfs_can_send_message(priv)))
+    {
+      canwarn("Failed to send CAN frame due to %s\n",
+              ret == CAN_INVALID_BUFFER ? "invalid buffer" :
+              ret == CAN_NO_MSG ? "no TX buffer available" : "unknown err");
+      return CAN_OK;
+    }
+
+  /* Increment statistics */
+
+  priv->stats.txb_sent++;
+  NETDEV_TXPACKETS(&priv->dev);
+
+#ifdef CONFIG_DEBUG_CAN_INFO
+  caninfo("Single CAN message transmit done\n");
+#endif
+
+  return CAN_OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_txpoll
+ *
+ * Description:
+ *   The transmitter is available, check if the network has any outgoing
+ *   packets ready to send.  This is a callback from devif_poll().
+ *   devif_poll() may be called:
+ *
+ *   1. When the preceding TX packet send is complete,
+ *   2. During normal TX polling
+ *
+ * Input Parameters:
+ *   dev  - Reference to the NuttX driver state structure
+ *
+ * Returned Value:
+ *   Zero (CAN_OK) on success; a negated errno on failure
+ *
+ * Assumptions:
+ *   May or may not be called from an interrupt handler.  In either case,
+ *   global interrupts are disabled, either explicitly or indirectly through
+ *   interrupt handling logic.
+ *
+ ****************************************************************************/
+
+static int mpfs_txpoll(struct net_driver_s *dev)
+{
+  mpfs_can_instance_t *priv =
+    (mpfs_can_instance_t *)dev->d_private;
+
+  /* If the polling resulted in data that should be sent out on the network,
+   * the field d_len is set to a value > 0.
+   */
+
+  if (priv->dev.d_len > 0)
+    {
+      /* Send the packet */
+
+      mpfs_transmit(priv);
+
+      /* Check if there is room in the device to hold another packet. If
+        * not, return a non-zero value to terminate the poll.
+        */
+
+      if (CAN_OK != mpfs_can_send_message_ready(priv))
+        {
+          return -EBUSY;
+        }
+    }
+
+  /* If zero is returned, the polling will continue until all connections
+   * have been examined.
+   */
+
+  return CAN_OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_txavail_work
+ *
+ * Description:
+ *   Perform an out-of-cycle poll on the worker thread.
+ *
+ * Input Parameters:
+ *   arg - Reference to the NuttX driver state structure (cast to void*)
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   Called on the higher priority worker thread.
+ *
+ ****************************************************************************/
+
+static void mpfs_txavail_work(void *arg)
+{
+  mpfs_can_instance_t *priv = (mpfs_can_instance_t *)arg;
+
+  /* Ignore the notification if the interface is not yet up */
+
+  if (priv->bifup)
```

Comment:
Is there a race between this and ifup/ifdown? Should actually this and also mpfs_can_send_message_ready be inside the net_lock? I am not sure, just checking

Reply:
yeah, both inside netlock is better since though slim but there is change that d_txavail is called bf ifup and af ifdown.